pkg/operator/customroute/custom_route_controller.go

```diff
@@ -0,0 +1,282 @@
+	configinformers "github.com/openshift/client-go/config/informers/externalversions/config/v1"
+	configlistersv1 "github.com/openshift/client-go/config/listers/config/v1"
+	routeclient "github.com/openshift/client-go/route/clientset/versioned/typed/route/v1"
+	routeinformer "github.com/openshift/client-go/route/informers/externalversions/route/v1"
+	routev1lister "github.com/openshift/client-go/route/listers/route/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/util/retry"
+
+	"github.com/openshift/library-go/pkg/certs"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resourcesynccontroller"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	"github.com/openshift/library-go/pkg/route/routecomponenthelpers"
+)
+
+type customRouteController struct {
+	destSecret     types.NamespacedName
+	componentRoute types.NamespacedName
+	ingressLister  configlistersv1.IngressLister
+	ingressClient  configsetterv1.IngressInterface
+	routeLister    routev1lister.RouteLister
+	routeClient    routeclient.RouteInterface
+	secretLister   corev1listers.SecretLister
+	resourceSyncer resourcesynccontroller.ResourceSyncer
+	operatorClient v1helpers.OperatorClient
+	targetRoute    *routev1.Route
+	consumingUsers []configv1.ConsumingUser
+}
+
+// RouteResourceGetter is a function pointer that is used by the client of this controller
+// to provide the Route template object. Typically, such an object is obtained using bindata but
+// the caller can also use Go Embed (or any other mechanism).
+type RouteResourceGetter func() *routev1.Route
+
+// NewCustomRouteController creates a new controller for custom CA for Routes.
+func NewCustomRouteController(
+	componentRouteNamespace string,
+	componentRouteName string,
+	destSecretNamespace string,
+	destSecretName string,
+	targetRouteResourceGetter RouteResourceGetter,
+	consumingUsers []configv1.ConsumingUser,
+	ingressInformer configinformers.IngressInformer,
+	ingressClient configsetterv1.IngressInterface,
+	routeInformer routeinformer.RouteInformer,
+	routeClient routeclient.RouteInterface,
+	kubeInformersForNamespaces v1helpers.KubeInformersForNamespaces,
+	operatorClient v1helpers.OperatorClient,
+	eventRecorder events.Recorder,
+	resourceSyncer resourcesynccontroller.ResourceSyncer,
+) factory.Controller {
+	controller := &customRouteController{
+		destSecret: types.NamespacedName{
+			Namespace: destSecretNamespace,
+			Name:      destSecretName,
+		},
+		componentRoute: types.NamespacedName{
+			Namespace: componentRouteNamespace,
+			Name:      componentRouteName,
+		},
+		targetRoute:    targetRouteResourceGetter(),
```

Comment:
I might have missed some discussion points and not a blocker for merging, but why not pass `*routev1.Route` directly?

Reply:
We went through a long path with @stlaz about this. I initially implement it this way. Then we switched to `bindata`. Finally, we switched into a function pointer, so that the caller can use either `bindata`, Go Embed or whatever he/she likes.

@stlaz wanted to keep it here to avoid side effects.

Reply:
leaving the final word to @stlaz but since the getter func is only used in the constructor I don't see a reason why we shouldn't pass the full reference and document that mutations happen in this controller or consequently use `AssetFunc` as is done in staticresourcecontroller.

Reply:
use a combination of the `AssetFunc` and `resourceread.ReadRouteV1OrDie`

Reply:
Done